pkg/pillar/hypervisor/kubevirt.go

```diff
@@ -0,0 +1,1355 @@
+			Name:        kubeName,
+			Namespace:   kubeapi.EVEKubeNameSpace,
+			Annotations: annotations,
+		},
+		Spec: k8sv1.PodSpec{
+			Containers: []k8sv1.Container{
+				{
+					Name:            kubeName,
+					Image:           ociName,
+					ImagePullPolicy: k8sv1.PullNever,
+					SecurityContext: &k8sv1.SecurityContext{
+						Privileged: &[]bool{true}[0],
+					},
+					Resources: k8sv1.ResourceRequirements{
+						Limits: k8sv1.ResourceList{
+							k8sv1.ResourceCPU:    resource.MustParse(vcpus),
+							k8sv1.ResourceMemory: resource.MustParse(memoryLimit),
+						},
+						Requests: k8sv1.ResourceList{
+							k8sv1.ResourceCPU:    resource.MustParse(vcpus),
+							k8sv1.ResourceMemory: resource.MustParse(memoryRequest),
+						},
+					},
+				},
+			},
+			DNSConfig: &k8sv1.PodDNSConfig{
+				Nameservers: []string{"8.8.8.8", "1.1.1.1"}, // XXX, temp, Add your desired DNS servers here
+			},
+		},
+	}
+	pod.Labels = make(map[string]string)
+	pod.Labels[eveLabelKey] = domainName
+	logrus.Infof("CreatePodConfig: pod setup %+v", pod)
+
+	// Now we have VirtualMachine Instance object, save it to config file for debug purposes
+	// and save it in context which will be used to start VM in Start() call
+	meta := vmiMetaData{
+		pod:      pod,
+		isPod:    true,
+		name:     kubeName,
+		domainID: int(rand.Uint32()),
+	}
+	ctx.vmiList[domainName] = &meta
+
+	podStr := fmt.Sprintf("%+v", pod)
+
+	// write to config file
+	file.WriteString(podStr)
+
+	return nil
+}
+
+func encodeSelections(selections []netattdefv1.NetworkSelectionElement) string {
+	bytes, err := json.Marshal(selections)
+	if err != nil {
+		logrus.Errorf("encodeSelections %v", err)
+		return ""
+	}
+	return string(bytes)
+}
+
+// StartPodContainer :  Starts container as kubernetes pod
+func StartPodContainer(kubeconfig *rest.Config, pod *k8sv1.Pod) error {
+
+	clientset, err := kubernetes.NewForConfig(kubeconfig)
+	if err != nil {
+		logrus.Errorf("StartPodContainer: can't get clientset %v", err)
+		return err
+	}
+
+	opStr := "created"
+	_, err = clientset.CoreV1().Pods(kubeapi.EVEKubeNameSpace).Create(context.TODO(), pod, metav1.CreateOptions{})
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			// TODO: update
+			logrus.Errorf("StartPodContainer: pod create filed: %v", err)
+			return err
+		} else {
+			opStr = "already exists"
+		}
+	}
+
+	logrus.Infof("StartPodContainer: Pod %s %s with nad %+v", pod.ObjectMeta.Name, opStr, pod.Annotations)
+
+	err = checkForPod(kubeconfig, pod.ObjectMeta.Name)
+	if err != nil {
+		logrus.Errorf("StartPodContainer: check for pod status error %v", err)
+		return err
+	}
+	logrus.Infof("StartPodContainer: Pod %s running", pod.ObjectMeta.Name)
+	return nil
+}
+
+func checkForPod(kubeconfig *rest.Config, podName string) error {
+	var i int
+	var status string
+	var err error
+	// wait for pod to be in running state, sometimes can take long, but we only wait for
+	// about a minute in order not to cause watchdog action
```

Comment:
See above. I assuming it is only the create and modify functions in domainmgr which call this, so it can wait forever (but you might want to check for any errors while waiting forever.)

Reply:
AFAIR, we retry here to make sure kubernetes is up and running after reboot, it could take more than 2 minutes or so.  The number of retries from domainmgr may not be sufficient, we do not want to generically bump up  retries in domainmgr.

Reply:
May be we can retry forever as long as error is not kubernetes not ready.

Reply:
I am sorry it's been longtime since we wrote this code.  We are looping for pod to be ready not k3s.  For k3s ready  we already do that in domainmgr when it starts.  We call kubeapi.WaitForKubernetes() that should ensure k3s is ready.

This additional loop check seems to be to make sure when we launched a pod it reached a ready status. It generally takes sometime for kubernetes to schedule the pod and start it. @naiming-zededa can you respond to this question.